src/qibolab/instruments/qm.py

```diff
@@ -581,7 +581,10 @@ def fetch_results(result, ro_pulses):
                 shots = handles.get(f"{serial}_shots").fetch_all().astype(int)
             else:
                 shots = None
-            results[pulse.qubit] = results[serial] = ExecutionResults.from_components(ires, qres, shots)
+            if average:
+                results[pulse.qubit] = results[serial] = AveragedResults.from_components(ires, qres)
+            else:
+                results[pulse.qubit] = results[serial] = ExecutionResults.from_components(ires, qres, shots)
```

Comment:
@stavros11 perhaps we can use a single if statement given that the one right above is related to shots.

Reply:
In this case, the `average` argument is only used to determine the output type. This looks artificial, since the exact same result is wrapped in a different type according to the user request (if it's just a recast, it can be done by the caller outside the function).

My opinion is that either is possible to determine `average` from something else, or we should just return `ExecutionResults` always.

(e.g. if `res` is an `ExecutionResults` object, you could recast just doing `AveragedResults(res.array)`, but we can even add a `AveragedResults.cast(res)` classmethod to make it more explicit)

Reply:
> In this case, the `average` argument is only used to determine the output type. This looks artificial, since the exact same result is wrapped in a different type according to the user request (if it's just a recast, it can be done by the caller outside the function).

I thought that this was the idea since `ExecutionResults` and `AveragedResults` are basically the same object.
I agree with the fact that we can probably move this call outside the sweeper since it is just a recasting.

Reply:
> Indeed it could be possible to infer from the returned results if an average occured or no. For example, one possibility is to check the shape of results and if the first (or last - we need to confirm this) axis has length `nshots` then averaging did not happen. But this check may fail if one uses a sweeper where the values happen to have length `nshots`, because it may mistakenly treat this axis as shots and think that averaging did not occur, while it really occured.

Indeed, I thought that the shape should change, but actually in an even more drastic way: not the `len` of one dimension to be different, but the `len` of the shape itself (so averaging or not should change the rank of the tensor).

But here I'm not an expert of the drivers, so you should tell me which are the actual outcomes.